file_writer_control/InThreadStatusTracker.py

```diff
@@ -0,0 +1,212 @@
+            new_command = CommandStatus(job_id, command_id)
+            self.known_commands[command_id] = new_command
+
+    def check_for_lost_connections(self):
+        """
+        Check workers, commands and jobs for the last update time and change the state of these if a timeout has been
+        reached.
+        """
+        now = datetime.now()
+        for worker in self.known_workers.values():
+            if (
+                worker.state != WorkerState.UNAVAILABLE
+                and now - worker.last_update > STATUS_MESSAGE_TIMEOUT
+            ):
+                worker.state = WorkerState.UNAVAILABLE
+        for command in self.known_commands.values():
+            if (
+                command.state != CommandState.SUCCESS
+                and command.state != CommandState.ERROR
+                and now - command.last_update > COMMAND_STATUS_TIMEOUT
+            ):
+                command.state = CommandState.TIMEOUT_RESPONSE
+        for job in self.known_jobs.values():
+            if (
+                job.state != JobState.DONE
+                and job.state != JobState.ERROR
+                and now - job.last_update > JOB_STATUS_TIMEOUT
+            ):
+                job.state = JobState.TIMEOUT
+
+    def process_answer(self, answer: Response):
+        """
+        Update workers, jobs and commands based on information in a response message.
+        :param answer: The response/answer message to use for status updates.
+        """
+        self.check_for_worker_presence(answer.service_id)
+        self.check_for_job_presence(answer.job_id)
+        self.check_for_command_presence(answer.job_id, answer.command_id)
+        new_job_state = extract_job_state_from_answer(answer)
+        if new_job_state is not None:
+            self.known_jobs[answer.job_id].state = new_job_state
+        self.known_commands[
+            answer.command_id
+        ].state = extract_state_from_command_answer(answer)
+        self.known_commands[answer.command_id].message = answer.message
+        self.known_jobs[answer.job_id].message = answer.message
+
+    def process_status(self, status_update: StatusMessage):
+        """
+        Update workers and jobs based on information in a status message.
+        :param status_update: The status message to use for updates.
+        """
+        self.check_for_worker_presence(status_update.service_id)
+        current_state = extract_worker_state_from_status(status_update)
+        self.known_workers[status_update.service_id].state = current_state
+        if current_state == WorkerState.WRITING:
+            job_id = json.loads(status_update.status_json)["job_id"]
+            self.check_for_job_presence(job_id)
+            self.known_jobs[job_id].state = JobState.WRITING
+            try:
+                self.known_jobs[job_id].service_id = status_update.service_id
+            except RuntimeError:
+                pass  # Expected error, do nothing
```

Comment:
I did not understand what is happening here.

Reply:
I have added some comments. Tell me if this aids in understanding what this statement does.